node/network/availability-distribution/src/lib.rs

```diff
@@ -807,96 +756,97 @@ where
 	}
 }
 
-/// Obtain all live candidates based on an iterator of relay heads.
-#[tracing::instrument(level = "trace", skip(ctx, relay_parents), fields(subsystem = LOG_TARGET))]
-async fn query_live_candidates_without_ancestors<Context>(
+#[derive(Debug)]
+enum FetchedLiveCandidate {
+	Cached,
+	Fresh(CandidateDescriptor),
+}
+
+/// Obtain all live candidates for all given `relay_blocks`.
+///
+/// This returns a set of all candidate hashes pending availability within the state
+/// of the explicitly referenced relay heads.
+///
+/// This also queries the provided `receipts` cache before reaching into the
+/// runtime and updates it with the information learned.
+#[tracing::instrument(level = "trace", skip(ctx, relay_blocks, receipts), fields(subsystem = LOG_TARGET))]
+async fn query_pending_availability_at<Context>(
 	ctx: &mut Context,
-	relay_parents: impl IntoIterator<Item = Hash>,
-) -> Result<HashSet<CommittedCandidateReceipt>>
+	relay_blocks: impl IntoIterator<Item = Hash>,
+	receipts: &mut HashMap<Hash, HashSet<CandidateHash>>,
+) -> Result<HashMap<CandidateHash, FetchedLiveCandidate>>
 where
 	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
 {
-	let iter = relay_parents.into_iter();
-	let hint = iter.size_hint();
+	let mut live_candidates = HashMap::new();
+
+	// fetch and fill out cache for each of these
+	for relay_parent in relay_blocks {
+		let receipts_for = match receipts.entry(relay_parent) {
+			Entry::Occupied(e) => {
+				live_candidates.extend(
```

Comment:
this clone is actually redundant: we own the `ccr` and we destruct it by taking out `descriptor`.

Reply:
probably addressed in #2134